File: basketball-csharp/BasketballTickets/Repository/Impl/MatchRepositoryImpl.cs
using BasketballTickets.Domain;
using BasketballTickets.Repository.Abstracts;
using BasketballTickets.Repository.Interfaces;
using BasketballTickets.Utils;
using Microsoft.Data.Sqlite;

namespace BasketballTickets.Repository.Impl;

public class MatchRepositoryImpl : AbstractDBRepository<long, Match>, IMatchRepository
{
    public MatchRepositoryImpl(DatabaseConfig dbConfig) : base(dbConfig) { }

    protected override string GetTableName() => "matches";

    protected override string GetInsertSql() =>
        "INSERT INTO matches (name, ticketPrice, totalSeats, availableSeats) VALUES (@name, @price, @total, @available)";

    protected override string GetUpdateSql() =>
        "UPDATE matches SET name=@name, ticketPrice=@price, totalSeats=@total, availableSeats=@available WHERE id=@id";

    protected override Match ExtractEntity(SqliteDataReader r) =>
        new(r.GetInt64(r.GetOrdinal("id")),
            r.GetString(r.GetOrdinal("name")),
            r.GetDouble(r.GetOrdinal("ticketPrice")),
            r.GetInt32(r.GetOrdinal("totalSeats")),
            r.GetInt32(r.GetOrdinal("availableSeats")));

    protected override void BindInsert(SqliteCommand cmd, Match m)
    {
        cmd.Parameters.AddWithValue("@name", m.Name);
        cmd.Parameters.AddWithValue("@price", m.TicketPrice);
        cmd.Parameters.AddWithValue("@total", m.TotalSeats);
        cmd.Parameters.AddWithValue("@available", m.AvailableSeats);
    }

    protected override void BindUpdate(SqliteCommand cmd, Match m)
    {
        cmd.Parameters.AddWithValue("@name", m.Name);
        cmd.Parameters.AddWithValue("@price", m.TicketPrice);
        cmd.Parameters.AddWithValue("@total", m.TotalSeats);
        cmd.Parameters.AddWithValue("@available", m.AvailableSeats);
        cmd.Parameters.AddWithValue("@id", m.Id);
    }

    protected override void SetGeneratedId(Match entity, long id) => entity.Id = id;

    // ---- Metode specifice domeniului ----

    public IList<Match> FindAvailableMatches()
    {
        Logger.Debug("FindAvailableMatches()");
        return QueryList("SELECT * FROM matches WHERE availableSeats > 0 ORDER BY id");
    }

    public IList<Match> FindSoldOutMatches()
    {
        Logger.Debug("FindSoldOutMatches()");
        return QueryList("SELECT * FROM matches WHERE availableSeats <= 0 ORDER BY id");
    }

    public Match UpdateAvailableSeats(long matchId, int seatsDelta)
    {
        Logger.Debug($"UpdateAvailableSeats(matchId={matchId}, delta={seatsDelta})");
        const string sql = "UPDATE matches SET availableSeats = availableSeats + @delta WHERE id=@id";

        using var conn = DbConfig.GetConnection();
        using var cmd = new SqliteCommand(sql, conn);
        cmd.Parameters.AddWithValue("@delta", seatsDelta);
        cmd.Parameters.AddWithValue("@id", matchId);
        cmd.ExecuteNonQuery();

        Logger.Info($"AvailableSeats updated for matchId={matchId} by {seatsDelta}");

        return FindById(matchId)
            ?? throw new InvalidOperationException($"Match not found after update: {matchId}");
    }

    // ---- Helper ----
    private List<Match> QueryList(string sql)
    {
        var results = new List<Match>();
        using var conn = DbConfig.GetConnection();
        using var cmd = new SqliteCommand(sql, conn);
        using var reader = cmd.ExecuteReader();
        while (reader.Read()) results.Add(ExtractEntity(reader));
        return results;
    }
}

File: java/src/main/java/org/example/repository/impl/MatchRepositoryImpl.java
package org.example.repository.impl;

import org.example.domain.Match;
import org.example.repository.abstracts.AbstractDBRepository;
import org.example.repository.interfaces.MatchRepository;
import org.example.utils.DatabaseConfig;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;


public class MatchRepositoryImpl
        extends AbstractDBRepository<Long, Match>
        implements MatchRepository {

    public MatchRepositoryImpl(DatabaseConfig dbConfig) {
        super(dbConfig);
    }

    @Override
    protected String getTableName() { return "matches"; }

    @Override
    protected String getInsertSql() {
        return "INSERT INTO matches (name, ticketPrice, totalSeats, availableSeats) VALUES (?, ?, ?, ?)";
    }

    @Override
    protected String getUpdateSql() {
        return "UPDATE matches SET name=?, ticketPrice=?, totalSeats=?, availableSeats=? WHERE id=?";
    }

    @Override
    protected Match extractEntity(ResultSet rs) throws SQLException {
        return new Match(
                rs.getLong("id"),
                rs.getString("name"),
                rs.getDouble("ticketPrice"),
                rs.getInt("totalSeats"),
                rs.getInt("availableSeats")
        );
    }

    @Override
    protected void bindInsert(PreparedStatement ps, Match m) throws SQLException {
        ps.setString(1, m.getName());
        ps.setDouble(2, m.getTicketPrice());
        ps.setInt(3, m.getTotalSeats());
        ps.setInt(4, m.getAvailableSeats());
    }

    @Override
    protected void bindUpdate(PreparedStatement ps, Match m) throws SQLException {
        ps.setString(1, m.getName());
        ps.setDouble(2, m.getTicketPrice());
        ps.setInt(3, m.getTotalSeats());
        ps.setInt(4, m.getAvailableSeats());
        ps.setLong(5, m.getId());
    }

    @Override
    protected void setGeneratedId(Match entity, long id) { entity.setId(id); }

    // ---- Domain-specific queries ----

    @Override
    public List<Match> findAvailableMatches() {
        logger.debug("findAvailableMatches()");
        return queryList("SELECT * FROM matches WHERE availableSeats > 0 ORDER BY id");
    }

    @Override
    public List<Match> findSoldOutMatches() {
        logger.debug("findSoldOutMatches()");
        return queryList("SELECT * FROM matches WHERE availableSeats <= 0 ORDER BY id");
    }

    @Override
    public Match updateAvailableSeats(Long matchId, int seatsDelta) {
        logger.debug("updateAvailableSeats(matchId={}, delta={})", matchId, seatsDelta);
        String sql = "UPDATE matches SET availableSeats = availableSeats + ? WHERE id=?";
        try (Connection conn = dbConfig.getConnection();
             PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, seatsDelta);
            ps.setLong(2, matchId);
            ps.executeUpdate();
            logger.info("availableSeats updated for matchId={} by {}", matchId, seatsDelta);
        } catch (SQLException e) {
            logger.error("updateAvailableSeats failed: {}", e.getMessage(), e);
            throw new RuntimeException("updateAvailableSeats failed", e);
        }
        return findById(matchId).orElseThrow();
    }

    // ---- Helper ----
    private List<Match> queryList(String sql) {
        List<Match> result = new ArrayList<>();
        try (Connection conn = dbConfig.getConnection();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {
            while (rs.next()) result.add(extractEntity(rs));
        } catch (SQLException e) {
            logger.error("queryList failed: {}", e.getMessage(), e);
        }
        return result;
    }
}